src/lib/hash/sha3/sha3.h

```diff
@@ -26,73 +30,9 @@ class SHA_3 : public HashFunction
       *                    224, 256, 384, or 512
       */
       explicit SHA_3(size_t output_bits);
-
-      size_t hash_block_size() const override { return m_bitrate / 8; }
-      size_t output_length() const override { return m_output_bits / 8; }
-
-      std::unique_ptr<HashFunction> new_object() const override;
-      std::unique_ptr<HashFunction> copy_state() const override;
```

Comment:
I believe those should have overloads in this class. It's somewhat hidden behind the polymorphy of `HashFunction` but as a user I would expect code along those lines to work:

```C++
SHA_3 sha(256);
HashFunction* other_sha = sha.new_object().get();

// this will fail, because `other_sha` is actually an instance of `Keccak_FIPS`
// and not of `SHA_3` (which I would have expected)
SHA_3* other_sha_ptr = dynamic_cast<SHA_3>(other_sha);
```

Reply:
If I understand correctly these overloads would be needed in all 4 base classes SHA_3_224, ... Can you confirm this?

Reply:
Yes, indeed. So that was actually an issue with the existing code already, no?
Come to think of it: `sha3.h` is an internal header, so applications don't actually have access to the `SHA_3_*` types and wouldn't be able to do the `dynamic_cast` as I suggested.

Still it is a code smell, in my opinion. @randombit what's your take on that?

Maybe [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) could be helpful here, if the `SHA_3` base class is not meant to be used as an interface by consuming code. It could potentially even be applied to `Keccak_FIPS` at one point.

A sketch for reference:

```C++
template <class DerivedT>
class SHA_3 : public Keccak_FIPS_generic
   {
   public:
      SHA_3(size_t output_bits) : Keccak_FIPS_generic(...) {}

      std::unique_ptr<HashFunction> new_object() const override
         {
         return std::make_unique<DerivedT>(...);
         }

      std::unique_ptr<HashFunction> copy_state() const override
         {
         return std::make_unique<DerivedT>(...);
         }
   };

class SHA_3_224 : public SHA_3<SHA_3_224> {}
class SHA_3_256 : public SHA_3<SHA_3_256> {}
// ...
```

Reply:
This seems like enourmously complicating vs just using ownership not inheritance.

SHA-3 is a hash function.
SHA-3 is not "a Keccak"

Reply:
This issue is unrelated to the inheritance of `Keccak`, though.

My mental model of `::new_object()` is that it returns an instance *of the very same hash function class*. The fact that `SHA_3` is not marked `final` but has sub-classes (`SHA_3_224`, `SHA_3_256`, ...) violates that mental model in the current implementation.

It's really just a nitpick, but it struck me as unusual. Internal code might do something like that:

```C++
std::unique_ptr<HashFunction> sha3_256 = std::make_unique<SHA_3_256>();
std::unique_ptr<HashFunction> bare_sha3 = sha3_256->new_object();
```

Now, both objects will behave exactly the same. But they won't be instances *of exactly the same class*. While `sha3_256` is an instance of the class `SHA_3_256`, `sha3_bare` will be an instance of `SHA_3`.

Applications will always go through `HashFunction::create("SHA-3(256)")` and therefore get an instance of `SHA_3` (without the `_256` suffix).

Again: just a minor code smell, if at all.